### PR TITLE
Docs for writing an application from scratch

### PR DESCRIPTION
**Pull request by tradej**
Resolves RHOARDOC-1148

- [x] Have you followed the guidelines in our [Contributing](http://wildfly-swarm.io/community/contributing/) document?
- [x] Have you created a [JIRA](https://issues.jboss.org/browse/SWARM) and used it in the commit message?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/wildfly-swarm/wildfly-swarm/pulls) for the same issue?
- [x] Have you built the project locally prior to submission with `mvn clean install`?

-----

This is the legacy content modularised. I have tried the procedure, and it works.

### CODE REVIEW

File: docs/howto/writing-an-application-from-scratch/src/main/resources/MyApp.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
+package com.example;
```
Comment (kenfinnigan) [inline docs/howto/writing-an-application-from-scratch/src/main/resources/MyApp.java]: Not sure this file is meant to be here?

File: docs/howto/writing-an-application-from-scratch/index.adoc [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+include::src/main/resources/pom.xml[]
+----
+:parameter-bom!:
+--
+
+. Create a directory structure for your application:
++
+[source,bash,options="nowrap"]
+----
+mkdir src/main/java
```
Comment (kenfinnigan) [inline docs/howto/writing-an-application-from-scratch/index.adoc]: I think you also need to create the directories for the package: com/example/rest

File: docs/howto/writing-an-application-from-scratch/index.adoc [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+After finishing the procedure, there should be a directory on your hard drive with the following contents:
+
+[source,options="nowrap"]
+----
+MyApp
+├── pom.xml
+└── src
+    └── main
+        └── java
```
Comment (kenfinnigan) [inline docs/howto/writing-an-application-from-scratch/index.adoc]: Needs to be updated to include package hierarchy

File: docs/howto/writing-an-application-from-scratch/index.adoc [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+. Create a directory for the application:
++
+--
+[source,bash,options="nowrap"]
+----
+$ mkdir myApp
+$ cd myApp
+----
+
+We recommend you start tracking the directory contents with Git.
```
Comment (Ladicek) [inline docs/howto/writing-an-application-from-scratch/index.adoc:25]: Maybe we should show `git init` then? :-)

Comment (Ladicek) [reply docs/howto/writing-an-application-from-scratch/index.adoc:25]: But that might lead us to showing things like `git commit` etc. etc., so probably a good idea to not dive into it too much.

File: docs/howto/writing-an-application-from-scratch/index.adoc [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+--
+
+. In the directory, create a `pom.xml` file with the following content.
++
+--
+// Product uses "bom-all" instead of "bom"
```
Comment (Ladicek) [inline docs/howto/writing-an-application-from-scratch/index.adoc]: It's the exact opposite -- community uses `bom-all` while product uses `bom`.

EDIT: Actually community can use `bom` too, and would be perfectly enough for this example.

File: docs/howto/writing-an-application-from-scratch/index.adoc [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+[source]
+----
+Hello from WildFly Swarm!
+----
+
+After finishing the procedure, there should be a directory on your hard drive with the following contents:
+
+[source,options="nowrap"]
+----
+MyApp
```
Comment (Ladicek) [inline docs/howto/writing-an-application-from-scratch/index.adoc]: Is it `myApp`?

File: docs/howto/writing-an-application-from-scratch/src/main/resources/HelloWorldEndpoint.java [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+import javax.ws.rs.GET;
+import javax.ws.rs.Produces;
+
+@Path("/hello")
+public class HelloWorldEndpoint {
+
+  @GET
+  @Produces("text/plain")
+  public Response doGet() {
+    return Response.ok("Hello from WildFly Swarm!").build();
```
Comment (Ladicek) [inline docs/howto/writing-an-application-from-scratch/src/main/resources/HelloWorldEndpoint.java:14]: Not sure how far we want to get in this simple example, but you can also just return a string:

```java
@GET
@Produces("text/plain")
public String get() {
    return "Hello from WildFly Swarm!";
}

File: docs/howto/writing-an-application-from-scratch/src/main/resources/pom.xml [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+
+  <dependencies>
+    <!-- Java EE 7 dependency -->
+    <dependency>
+      <groupId>javax</groupId>
+      <artifactId>javaee-api</artifactId>
+      <version>7.0</version>
+      <scope>provided</scope>
+    </dependency>
+    <!-- Wildfly Swarm Fractions -->
```
Comment (Ladicek) [inline docs/howto/writing-an-application-from-scratch/src/main/resources/pom.xml]: You typically want _either_ Java EE 7 APIs, _or_ WildFly Swarm fractions. If you include the fraction dependencies, you don't have to include the Java EE APIs, as the fractions bring them transitively.

Now, a question: what do we want to promote? Depending on Java EE APIs and fraction autodetection, or depending on fractions directly?

My personal opinion is that we want to promote depending directly on the fractions, in which case I'd remove the Java EE API dependency and added a dependency on `org.wildfly.swarm:jaxrs`.

File: docs/howto/writing-an-application-from-scratch/src/main/resources/pom.xml [hunk 9]
```diff
@@ -0,0 +1,61 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+    xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
+
+  <modelVersion>4.0.0</modelVersion>
+  <groupId>com.example</groupId>
+  <artifactId>restful-endpoint</artifactId>
+  <name>Wildfly Swarm Example</name>
+  <version>1.0.0-SNAPSHOT</version>
+  <packaging>war</packaging>
```
Comment (Ladicek) [inline docs/howto/writing-an-application-from-scratch/src/main/resources/pom.xml]: By convention (https://maven.apache.org/developers/conventions/code.html#POM_Code_Convention), this block should look more like

```xml
<modelVersion>4.0.0</modelVersion>

<groupId>com.example</groupId>
<artifactId>restful-endpoint</artifactId>
<version>1.0.0-SNAPSHOT</version>
<packaging>war</packaging>

<name>Wildfly Swarm Example</name>
```

File: docs/howto/writing-an-application-from-scratch/src/main/resources/pom.xml [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+        <groupId>org.wildfly.swarm</groupId>
+        <artifactId>{parameter-bom}</artifactId>
+        <version>${version.wildfly.swarm}</version>
+        <scope>import</scope>
+        <type>pom</type>
+      </dependency>
+    </dependencies>
+  </dependencyManagement>
+
+  <build>
```
Comment (Ladicek) [inline docs/howto/writing-an-application-from-scratch/src/main/resources/pom.xml]: By convention (https://maven.apache.org/developers/conventions/code.html#POM_Code_Convention), the `<build>` block should be after `<dependencies>`.

File: docs/howto/writing-an-application-from-scratch/src/main/resources/pom.xml [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+    <dependency>
+      <groupId>javax</groupId>
+      <artifactId>javaee-api</artifactId>
+      <version>7.0</version>
+      <scope>provided</scope>
+    </dependency>
+    <!-- Wildfly Swarm Fractions -->
+  </dependencies>
+</project>
+
```
Comment (Ladicek) [inline docs/howto/writing-an-application-from-scratch/src/main/resources/pom.xml]: I also see that @inoxx03 adds Fabric8 Maven plugin configuration for deploying to OpenShift (see https://github.com/fabric8-launcher/launcher-documentation/pull/1054). Do we maybe want to be consistent in this?

File: docs/howto/writing-an-application-from-scratch/index.adoc (hunk 4 above)
Comment (tradej) [reply docs/howto/writing-an-application-from-scratch/index.adoc:25]: Okay, I'll add that to Related information.

File: docs/howto/writing-an-application-from-scratch/index.adoc (hunk 5 above)
Comment (tradej) [reply docs/howto/writing-an-application-from-scratch/index.adoc]: OK, I'll use `bom` for simplicity then.

File: docs/howto/writing-an-application-from-scratch/src/main/resources/MyApp.java (hunk 1 above)
Comment (tradej) [reply docs/howto/writing-an-application-from-scratch/src/main/resources/MyApp.java]: Of course not, I shouldn't `git add` entire directories. I also shouldn't be working that late. Thank you.

File: docs/howto/writing-an-application-from-scratch/src/main/resources/pom.xml (hunk 9 above)
Comment (tradej) [reply docs/howto/writing-an-application-from-scratch/src/main/resources/pom.xml]: Thanks, will do.

File: docs/howto/writing-an-application-from-scratch/src/main/resources/pom.xml (hunk 10 above)
Comment (tradej) [reply docs/howto/writing-an-application-from-scratch/src/main/resources/pom.xml]: Thanks, will do.

File: docs/howto/writing-an-application-from-scratch/index.adoc (hunk 2 above)
Comment (tradej) [reply docs/howto/writing-an-application-from-scratch/index.adoc]: Ah, of course.

File: docs/howto/writing-an-application-from-scratch/src/main/resources/pom.xml (hunk 11 above)
Comment (tradej) [reply docs/howto/writing-an-application-from-scratch/src/main/resources/pom.xml]: Certainly, I'll add that, thank you.

**Comment by tradej** [comment]
Please review the changes, I'll squash the commits before merge.

**Review by Ladicek — DISMISSED** [review]

**Comment by tradej** [comment]
Squashed. Please merge and cherry-pick to the `7.1.x` branch in the prod repository.

**Review by kenfinnigan — APPROVED** [review]

**Pull request merged**